# JSON examples and exercise
****
+ get familiar with packages for dealing with JSON
+ study examples with JSON strings and files 
+ work on exercise to be completed and submitted 
****
+ reference: http://pandas.pydata.org/pandas-docs/stable/io.html#io-json-reader
+ data source: http://jsonstudio.com/resources/
****

In [199]:
import pandas as pd

## imports for Python, Pandas

In [200]:
import json
from pandas.io.json import json_normalize

## JSON example, with string

+ demonstrates creation of normalized dataframes (tables) from nested json string
+ source: http://pandas.pydata.org/pandas-docs/stable/io.html#normalization

In [204]:
# define json string
data = [{'state': 'Florida', 
         'shortname': 'FL',
         'info': {'governor': 'Rick Scott'},
         'counties': [{'name': 'Dade', 'population': 12345},
                      {'name': 'Broward', 'population': 40000},
                      {'name': 'Palm Beach', 'population': 60000}]},
        {'state': 'Ohio',
         'shortname': 'OH',
         'info': {'governor': 'John Kasich'},
         'counties': [{'name': 'Summit', 'population': 1234},
                      {'name': 'Cuyahoga', 'population': 1337}]}]

In [205]:
# use normalization to create tables from nested element
json_normalize(data, 'counties')

name  population
0        Dade       12345
1     Broward       40000
2  Palm Beach       60000
3      Summit        1234
4    Cuyahoga        1337

In [206]:
# further populate tables created from nested element
json_normalize(data, 'counties', ['state', 'shortname', ['info', 'governor']])

name  population    state shortname info.governor
0        Dade       12345  Florida        FL    Rick Scott
1     Broward       40000  Florida        FL    Rick Scott
2  Palm Beach       60000  Florida        FL    Rick Scott
3      Summit        1234     Ohio        OH   John Kasich
4    Cuyahoga        1337     Ohio        OH   John Kasich

****
## JSON example, with file

+ demonstrates reading in a json file as a string and as a table
+ uses small sample file containing data about projects funded by the World Bank 
+ data source: http://jsonstudio.com/resources/

In [207]:
# load json as string
json.load((open('data/world_bank_projects_less.json')))

[{'_id': {'$oid': '52b213b38594d8a2be17c780'},
  'approvalfy': 1999,
  'board_approval_month': 'November',
  'boardapprovaldate': '2013-11-12T00:00:00Z',
  'borrower': 'FEDERAL DEMOCRATIC REPUBLIC OF ETHIOPIA',
  'closingdate': '2018-07-07T00:00:00Z',
  'country_namecode': 'Federal Democratic Republic of Ethiopia!$!ET',
  'countrycode': 'ET',
  'countryname': 'Federal Democratic Republic of Ethiopia',
  'countryshortname': 'Ethiopia',
  'docty': 'Project Information Document,Indigenous Peoples Plan,Project Information Document',
  'envassesmentcategorycode': 'C',
  'grantamt': 0,
  'ibrdcommamt': 0,
  'id': 'P129828',
  'idacommamt': 130000000,
  'impagency': 'MINISTRY OF EDUCATION',
  'lendinginstr': 'Investment Project Financing',
  'lendinginstrtype': 'IN',
  'lendprojectcost': 550000000,
  'majorsector_percent': [{'Name': 'Education', 'Percent': 46},
   {'Name': 'Education', 'Percent': 26},
   {'Name': 'Public Administration, Law, and Justice', 'Percent': 16},
   {'Name': 'Educatio

In [208]:
# load as Pandas dataframe
#sample_json_df = pd.read_json('data/world_bank_projects_less.json')
#sample_json_df
sample_json_df.columns
#sample_json_df['mjtheme']
#sample_json_df['countryname']

Index(['_id', 'approvalfy', 'board_approval_month', 'boardapprovaldate',
       'borrower', 'closingdate', 'country_namecode', 'countrycode',
       'countryname', 'countryshortname', 'docty', 'envassesmentcategorycode',
       'grantamt', 'ibrdcommamt', 'id', 'idacommamt', 'impagency',
       'lendinginstr', 'lendinginstrtype', 'lendprojectcost',
       'majorsector_percent', 'mjsector_namecode', 'mjtheme',
       'mjtheme_namecode', 'mjthemecode', 'prodline', 'prodlinetext',
       'productlinetype', 'project_abstract', 'project_name', 'projectdocs',
       'projectfinancialtype', 'projectstatusdisplay', 'regionname', 'sector',
       'sector1', 'sector2', 'sector3', 'sector4', 'sector_namecode',
       'sectorcode', 'source', 'status', 'supplementprojectflg', 'theme1',
       'theme_namecode', 'themecode', 'totalamt', 'totalcommamt', 'url'],
      dtype='object')

In [209]:
data1 = json.load((open('data/world_bank_projects_less.json')))

df1 = json_normalize(data1, 'mjtheme_namecode')
df1 = df1.set_index('code')
#df1
#df1.index
#df1.values
#df1.loc['11','name']

df2 = pd.read_json('data/world_bank_projects_less.json')
df2[['countryname']]

countryname
0  Federal Democratic Republic of Ethiopia
1                      Republic of Tunisia

****
## JSON exercise

Using data in file 'data/world_bank_projects.json' and the techniques demonstrated above,
1. Find the 10 countries with most projects
2. Find the top 10 major project themes (using column 'mjtheme_namecode')
3. In 2. above you will notice that some entries have only the code and the name is missing. Create a dataframe with the missing names filled in.

# Question 1

In [210]:
# load the file 'world_bank_projects.json' as Pandas dataframe
json_df = pd.read_json('data/world_bank_projects.json')

json_df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 500 entries, 0 to 499
Data columns (total 50 columns):
_id                         500 non-null object
approvalfy                  500 non-null int64
board_approval_month        500 non-null object
boardapprovaldate           500 non-null object
borrower                    485 non-null object
closingdate                 370 non-null object
country_namecode            500 non-null object
countrycode                 500 non-null object
countryname                 500 non-null object
countryshortname            500 non-null object
docty                       446 non-null object
envassesmentcategorycode    430 non-null object
grantamt                    500 non-null int64
ibrdcommamt                 500 non-null int64
id                          500 non-null object
idacommamt                  500 non-null int64
impagency                   472 non-null object
lendinginstr                495 non-null object
lendinginstrtype            495 non

 The file has 500 rows and 50 columns

In [211]:
json_df[['countryname']]

countryname
0     Federal Democratic Republic of Ethiopia
1                         Republic of Tunisia
2                                      Tuvalu
3                           Republic of Yemen
4                          Kingdom of Lesotho
5                           Republic of Kenya
6                           Republic of India
7                  People's Republic of China
8                           Republic of India
9                          Kingdom of Morocco
10                    Republic of South Sudan
11                          Republic of India
12                          Republic of Ghana
13         Democratic Republic of Timor-Leste
14                Hashemite Kingdom of Jordan
15                                      Samoa
16                                      Samoa
17                     Republic of Madagascar
18                        Kingdom of Cambodia
19                         Kingdom of Morocco
20                            Kyrgyz Republic
21                                      Nepal
22                Hashemite Kingdom of Jordan
23                     Republic of Tajikistan
24                     Republic of Azerbaijan
25                      East Asia and Pacific
26           Lao People's Democratic Republic
27                            Pacific Islands
28                            Solomon Islands
29                     Republic of Mozambique
..                                        ...
470                       Republic of Liberia
471           People's Republic of Bangladesh
472                       Republic of Armenia
473                       Republic of Albania
474              Islamic Republic of Pakistan
475                     Republic of Indonesia
476             Socialist Republic of Vietnam
477                    Republic of Uzbekistan
478              Islamic Republic of Pakistan
479                     Republic of Guatemala
480                        Republic of Zambia
481                         Republic of India
482                         Republic of India
483                         Republic of India
484               Republic of the Philippines
485                                  Mongolia
486              Middle East and North Africa
487                     United Mexican States
488             Federative Republic of Brazil
489  The Independent State of Papua New Guine
490                                     Nepal
491                        West Bank and Gaza
492              Islamic Republic of Pakistan
493           People's Republic of Bangladesh
494                                     Nepal
495                                   Jamaica
496          Lao People's Democratic Republic
497                        Republic of Guinea
498                     Republic of Indonesia
499                         Republic of Kenya

[500 rows x 1 columns]

The column 'countryname' has 500 entries and none of them are missing, 
so we count the values of this column, and return top 10 countries

In [212]:
json_df['countryname'].value_counts()[:10]

Republic of Indonesia              19
People's Republic of China         19
Socialist Republic of Vietnam      17
Republic of India                  16
Republic of Yemen                  13
People's Republic of Bangladesh    12
Nepal                              12
Kingdom of Morocco                 12
Republic of Mozambique             11
Africa                             11
Name: countryname, dtype: int64

In [ ]:
Africa is clearly not a country, so this is probably a data error.

# Question 2

We Load the file as a string, use the json_normalize function to extract 
the column 'mjtheme_namecode', and convert it to its own data frame

In [256]:
import numpy as np

data2 = json.load((open('data/world_bank_projects.json')))
df_mjtheme_namecode = json_normalize(data2, 'mjtheme_namecode')
df_mjtheme_namecode.tail(20)
df_mjtheme_namecode.head(20)



code                                          name
0     8                             Human development
1    11                                              
2     1                           Economic management
3     6         Social protection and risk management
4     5                         Trade and integration
5     2                      Public sector governance
6    11  Environment and natural resources management
7     6         Social protection and risk management
8     7                   Social dev/gender/inclusion
9     7                   Social dev/gender/inclusion
10    5                         Trade and integration
11    4      Financial and private sector development
12    6         Social protection and risk management
13    6                                              
14    2                      Public sector governance
15    4      Financial and private sector development
16   11  Environment and natural resources management
17    8                                              
18   10                             Rural development
19    7

We see that each theme has its own unique code ranging from 1 to 11. However, some themes 
corresponding to the various codes are missing from the name 'column. This means that we must 
count by the values in the 'code' column rather than by the values in the 'name' column.

In [251]:
# Let's count the values of the themes
print(df_mjtheme_namecode['name'].value_counts())

Environment and natural resources management    223
Rural development                               202
Human development                               197
Public sector governance                        184
Social protection and risk management           158
Financial and private sector development        130
                                                122
Social dev/gender/inclusion                     119
Trade and integration                            72
Urban development                                47
Economic management                              33
Rule of law                                      12
Name: name, dtype: int64


One can see that there are 122 missing values, meaning that we can not count by 'name' column. 
Instead, we first create the list of codes corresponding to the top 10 major project themes, sorting
them in descending order, and only then print the names of themes corresponding to these codes. In doing so, we replace '' with NaN's first, drop them subsequently, and apply unique method. 

In [260]:
list_of_codes = list(df_mjtheme_namecode['code'].value_counts()[:10].keys())

for code in list_of_codes:
    print(df_mjtheme_namecode.set_index('code').loc[code, 'name'].replace('', np.nan).dropna().unique()[0])
    

Environment and natural resources management
Rural development
Human development
Public sector governance
Social protection and risk management
Financial and private sector development
Social dev/gender/inclusion
Trade and integration
Urban development
Economic management


# Question 3 

To perform the task of this question, we first create the lists of codes and corresponding themes, then make the dictionary in which each code corresponds to the theme it represents. After that we use this dictionary to iterate over the dataframe replacing missing themes.

In [329]:

# Create the list of codes
list_of_codes = list(df_mjtheme_namecode['code'].value_counts().keys())

# Auxilliary data-frame with the rows corresponding to missing themes removed 
df_aux = df_mjtheme_namecode.replace('', np.nan).dropna()

# Create the list of themes
list_of_themes = []
for code in list_of_codes:
    list_of_themes.append(df_aux.set_index('code').loc[code, 'name'].unique()[0])
   
 # Create a dictionary in which each code corresponds to the theme it represents 
dictionary = {k: v for k,v in zip(list_of_codes, list_of_themes)}


# Replace the empty strings with the theme names corresponding to their codes.
for i in range(len(df_mjtheme_namecode.index)):
    if (df_mjtheme_namecode['name'][i] == ''):
        df_mjtheme_namecode['name'][i] = dictionary[df_mjtheme_namecode['code'][i]]


df_mjtheme_namecode.head(50)

code                                          name
0     8                             Human development
1    11  Environment and natural resources management
2     1                           Economic management
3     6         Social protection and risk management
4     5                         Trade and integration
5     2                      Public sector governance
6    11  Environment and natural resources management
7     6         Social protection and risk management
8     7                   Social dev/gender/inclusion
9     7                   Social dev/gender/inclusion
10    5                         Trade and integration
11    4      Financial and private sector development
12    6         Social protection and risk management
13    6         Social protection and risk management
14    2                      Public sector governance
15    4      Financial and private sector development
16   11  Environment and natural resources management
17    8                             Human development
18   10                             Rural development
19    7                   Social dev/gender/inclusion
20    2                      Public sector governance
21    2                      Public sector governance
22    2                      Public sector governance
23   10                             Rural development
24    2                      Public sector governance
25   10                             Rural development
26    6         Social protection and risk management
27    6         Social protection and risk management
28   11  Environment and natural resources management
29    4      Financial and private sector development
30    5                         Trade and integration
31   11  Environment and natural resources management
32   10                             Rural development
33    6         Social protection and risk management
34    6         Social protection and risk management
35    6         Social protection and risk management
36   10                             Rural development
37   11  Environment and natural resources management
38    5                         Trade and integration
39   10                             Rural development
40    2                      Public sector governance
41    8                             Human development
42    7                   Social dev/gender/inclusion
43    8                             Human development
44    8                             Human development
45    2                      Public sector governance
46    7                   Social dev/gender/inclusion
47    8                             Human development
48    2                      Public sector governance
49    2                      Public sector governance